src/main/java/graphql/GraphQL.java

```diff
@@ -479,13 +480,18 @@ public ExecutionResult execute(ExecutionInput executionInput) {
 
 
     private CompletableFuture<ExecutionResult> parseValidateAndExecute(ExecutionInput executionInput, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {
-        PreparsedDocumentEntry preparsedDoc = preparsedDocumentProvider.get(executionInput.getQuery(), query -> parseAndValidate(executionInput, graphQLSchema, instrumentationState));
-
+        AtomicReference<ExecutionInput> executionInputRef = new AtomicReference<>(executionInput);
```

Comment:
No - its a lambda changing the outside variable.  So it has to be final or effectively final.

Using AtomicRef allows you to bend this rule.  That is I can call the lambda and allow it to have side effects.

Ideally the code would have been structure so that its free of side effects  (calling all the way down) but it wasn't and I judged a complete refactor as too much.

Reply:
ah right ... all good ... the bit hacky alternative would be `new ExecutionInput[]{executionInput}`, but atomic reference is definitely nicer